azimuth_schedule_operator/models/v1alpha1/lease.py

```diff
@@ -0,0 +1,127 @@
+import datetime as dt
+import typing as t
+
+from pydantic import Field
+
+from kube_custom_resource import CustomResource, schema
+
+
+class Machine(schema.BaseModel):
+    """Represents a reservation for a machine."""
+
+    size_id: schema.constr(min_length=1) = Field(
+        ..., description="The ID of the size for the machine."
+    )
+    count: schema.conint(gt=0) = Field(
+        ..., description="The number of machines of this size to reserve."
+    )
+
+
+class ResourcesSpec(schema.BaseModel):
+    """The resources that a lease is reserving."""
+
+    machines: t.List[Machine] = Field(
+        default_factory=list,
+        description="Machines that should be reserved by the lease.",
+    )
+
+
+class LeaseSpec(schema.BaseModel):
+    """The spec of a lease."""
+
+    cloud_credentials_secret_name: schema.constr(min_length=1) = Field(
+        ..., description="The name of the secret containing the cloud credentials."
+    )
+    starts_at: schema.Optional[dt.datetime] = Field(
+        None,
+        description=(
+            "The start time for the lease. "
+            "If no start time is given, it is assumed to start immediately."
+        ),
+    )
+    ends_at: schema.Optional[dt.datetime] = Field(
+        None,
+        description=(
+            "The end time for the lease. "
+            "If no end time is given, the lease is assumed to be infinite."
+        ),
+    )
+    grace_period: schema.Optional[schema.conint(ge=0)] = Field(
+        None,
+        description=(
+            "The grace period before the end of the lease that the platform "
+            "will be given to shut down gracefully. "
+            "If not given, the operator default grace period will be used."
+        ),
+    )
+    resources: ResourcesSpec = Field(
+        ..., description="The resources that the lease is reserving."
+    )
+
+
+class LeasePhase(str, schema.Enum):
+    """The phase of a lease."""
+
+    # Stable phases
+    PENDING = "Pending"
+    ACTIVE = "Active"
+    TERMINATED = "Terminated"
+    ERROR = "Error"
+    # Transitional phases
+    CREATING = "Creating"
+    STARTING = "Starting"
+    UPDATING = "Updating"
+    TERMINATING = "Terminating"
+    DELETING = "Deleting"
```

Comment:
I am not sure we need to report all these phases in the Lease Phases, but I guess other ways of doing this are simply more complicated.